rust-symcrypt/src/gcm.rs

```diff
@@ -0,0 +1,409 @@
+/* Galois Counter Mode functions. For further documentation please refer to symcrypt.h */
+
+use crate::block_ciphers::*;
+use crate::errors::SymCryptError;
+use core::ffi::c_void;
+use std::mem;
+use std::ptr;
+use std::vec;
+use symcrypt_sys;
+
+pub struct GcmState {
+    expanded_key: symcrypt_sys::SYMCRYPT_GCM_EXPANDED_KEY,
+    state: symcrypt_sys::SYMCRYPT_GCM_STATE,
+}
+
+// Have to Box<> since memory address for Self is moved around when returning from GcmState::new()
+// My guess is that it has to do with Result<> wrapping the Self and SymCryptError.
+// The only accepted Cipher for GCM is AesBlock.
+impl GcmState {
+    pub fn new(
+        key: &[u8],
+        nonce: &[u8],
+        cipher: BlockCipherType,
+    ) -> Result<Box<Self>, SymCryptError> {
+        let mut instance = Box::new(GcmState {
+            state: symcrypt_sys::SYMCRYPT_GCM_STATE::default(),
+            expanded_key: symcrypt_sys::SYMCRYPT_GCM_EXPANDED_KEY::default(),
+        });
+
+        expand_key(key, &mut instance.expanded_key, convert_cipher(cipher))?;
+        unsafe {
+            // SAFETY: FFI calls
+            symcrypt_sys::SymCryptGcmInit(
+                &mut instance.state,
+                &instance.expanded_key,
+                nonce.as_ptr(),
+                nonce.len() as symcrypt_sys::SIZE_T,
+            );
+            Ok(instance)
+        }
+    }
+
+    pub fn auth(&mut self, auth_data: &[u8]) {
+        unsafe {
+            // SAFETY: FFI calls
+            symcrypt_sys::SymCryptGcmAuthPart(
+                &mut self.state,
+                auth_data.as_ptr(),
+                auth_data.len() as symcrypt_sys::SIZE_T,
+            );
+        }
+    }
+
+    pub fn encrypt_part(&mut self, data: &[u8]) -> Vec<u8> {
```

Comment:
This will potentially incur a lot of copying when encrypting large amounts of data, as you need to put all of the pieces together at the end. I know we've gone back and forth on how much to write the code for performance versus idiomatic Rust design, but I think the performance impact here could be a lot higher than in the other cases we've discussed because the copying will probably take vastly more time than the encryption itself.

Is there a way to take a slice of an existing array/vector as a mutable input and write the result to that?

If you want to leave it as-is for now and make a decision based on benchmarking it later, I'm fine with that. But I do expect that there will be significant performance impact.

Same comment for decrypt.

Reply:
Maybe add another function, such as fast_encrypt_part() that updates an output parameter instead of returning a result..

Reply:
We can take in an existing slice and modify it, similar to how `expand_key()` works. I am open to changing it over now, will track this as a discussion item

Reply:
Might also want to do for the chacha encrypt/decrypt functions.


The pattern for encrypt/decrypt would be to do as an out parameter.

Reply:
> Maybe add another function, such as fast_encrypt_part() that updates an output parameter instead of returning a result..

I dont see a point in doing a fast version, Inherently this is already supposed to be fast, if the speed is an issue I think we have to just move over to the out parameter.